whitepapers/Dates.md

```diff
@@ -0,0 +1,72 @@
+# Partially Unknown Dates
+
+## Overview
+Many works in SHARE have the wrong publication date displayed, because many sources provide only partial publication dates for some works, such as "May 2016" or just "2016".
+When the day of the month is omitted, the current date parsing logic assumes the first day of the month. When the month is omitted, it assumes January.
+
+### Goals
+Allow storing and displaying works with partially-known publication dates.
+
+## Changes
+
+Create a custom `PartialDate` class with:
+* Three main properties:
+    * `year`: integer (required)
+    * `month`: integer (optional)
+    * `day`: integer (optional)
+* Missing/null values indicate an unknown date part
+* Validation that set values are the correct type and correspond to a valid date
+    * If `month` is missing, `day` must also be missing
+* A way to construct a `PartialDate` from a `Date`-like object, string, or 8-digit integer (see below)
+
+### Models
+* Custom Django field `PartialDateField` that inherits from `IntegerField`
+    * Store a `PartialDate` instance as an 8-digit integer in the database, `YYYYMMDD`
+    * If `month` or `day` is null, put `00` in the `MM` or `DD` segment, respectively
+        * "February 26, 2016" stored as `20160226`
+        * "February 2016" stored as `20160200`
+        * "2016" stored as `20160000`
+    * This allows sorting and filtering by a `PartialDateField` in the database
+        * Values with unknown month and day are sorted at the beginning of the year
+        * Values with known month but unknown day are sorted at the beginning of the month
+* Make `AbstractCreativeWork.date_published` a nullable `PartialDateField`
+
+### Migrations
+* Change the `date_published` columns on `share_creativework` and `share_creativeworkversion` from `DATETIME` to `INTEGER`
+* Convert datetime values into integers, as above, dropping the "time" part of the datetime
+    * If a value is at exactly midnight on the first day of the month, assume unknown day
+        * If the month is January, also assume unknown month
+    * Examples:
+        * `2016-03-20 06:02:11` becomes `20160320`
+        * `2016-03-20 00:00:00` becomes `20160320`
+        * `2016-03-01 06:02:11` becomes `20160301`
+        * `2016-03-01 00:00:00` becomes `20160300`
+        * `2016-01-01 06:02:11` becomes `20160101`
+        * `2016-01-01 00:00:00` becomes `20160000`
+
+### Transformers
+* Add a `PartialDateParser` Transformer link that parses a string into a `PartialDate` instance
```

Comment:
[Arrow docs](http://arrow.readthedocs.io/en/latest/) have a nice table of common date formats:

Token | Output
------ | --------
YYYY | 2000, 2001, 2002 ... 2012, 2013
YY | 00, 01, 02 ... 12, 13
DDDD | 001, 002, 003 ... 364, 365
DDD | 1, 2, 3 ... 4, 5

Token | Output
------ | --------
MMMM | January, February, March ...
MMM | Jan, Feb, Mar ...
MM | 01, 02, 03 ... 11, 12
M | 1, 2, 3 ... 11, 12
DD | 01, 02, 03 ... 30, 31
D | 1, 2, 3 ... 30, 31
Do | 1st, 2nd, 3rd ... 30th, 31st

Token | Output
------ | --------
dddd | Monday, Tuesday, Wednesday ...
ddd | Mon, Tue, Wed ...
d | 1, 2, 3 ... 6, 7